### ISSUE TITLE
rocket::form::FromForm derive generates code that triggers a bunch of clippy warnings

### ISSUE THREAD

**Issue by haarts** [SergioBenitez/Rocket#1670]
Rocket version: dc5c57b624cebd52063c645e2de2e4cc978d960d
Operating system: PopOS 20.10

I want to accept an HTTP request with a query string. For this I created a struct:

```rust
#[derive(Clone, Debug, PartialEq, rocket::form::FromForm)]
pub struct Query<'r> {
    q: &'r str,
}
```
This works but clippy is deeply unhappy about it. No less that 5 lints:

```
error: used binding `_o` which is prefixed with an underscore. A leading underscore signals that a binding will not be used
  --> internal.rs:17:8
   |
17 |     q: &'r str,
   |        ^
   |
   = note: `-D clippy::used-underscore-binding` implied by `-D clippy::pedantic`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#used_underscore_binding

error: called `map(<f>).unwrap_or_else(<g>)` on an `Option` value. This can be done more directly by calling `map_or_else(<g>, <f>)` instead
  --> internal.rs:17:8
   |
17 |     q: &'r str,
   |        ^ help: try this: `q: &.map_or_else(&, &)`
   |
   = note: `-D clippy::map-unwrap-or` implied by `-D clippy::pedantic`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_unwrap_or

error: you seem to be trying to match on a boolean expression
  --> internal.rs:17:8
   |
17 |     q: &'r str,
   |        ^ help: consider using an `if`/`else` expression: `if & { rocket::form::FromForm } else { rocket::form::FromForm }`
   |
   = note: `-D clippy::match-bool` implied by `-D clippy::pedantic`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#match_bool

error: you seem to be trying to match on a boolean expression
  --> internal.rs:17:8
   |
17 |     q: &'r str,
   |        ^ help: consider using an `if`/`else` expression: `if & { rocket::form::FromForm } else { & }`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#match_bool

error: used binding `_e` which is prefixed with an underscore. A leading underscore signals that a binding will not be used
  --> internal.rs:15:22
   |
15 |   pub struct Query<'r> {
   |  ______________________^
16 | |     #[allow(clippy::used_underscore_binding, clippy::map_unwrap_or)]
17 | |     q: &'r str,
18 | | }
   | |_^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#used_underscore_binding

error: aborting due to 5 previous errors

```

The first one seems to be related to https://github.com/SergioBenitez/Rocket/issues/1656 and I believe is triggered by this line: https://github.com/SergioBenitez/Rocket/blob/master/core/codegen/src/derive/from_form.rs#L15

The remaining issues are a bit beyond my current skill set to debug. 

I'd be happy to suppress the errors but adding the relevant `allow` directives doesn't seem to work with generated code.

**Comment by haarts** [comment]
I've worked around the warnings by wrapping my struct in a mod and adding the allow directives there.

**Comment by SergioBenitez** [comment]
Oh clippy clippy. Okay. I have changes on top of a WIP branch that will resolve these. As soon as those changes are ready to commit, the other will come along for the ride.

**Comment by SergioBenitez** [comment]
Resolved in 1a42009e9f729661868d339c77f5b6fc8757cebe.
